{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Notes on building an automatic cell type classifier for scRNA-seq data\n",
    "Implementation of SVM and ANN to achieve cell type identifications based on previously annotated datasets.\n",
    "\n",
    "10-13-2020:\n",
    "- attempting to import data using ScanPy (https://scanpy.readthedocs.io/en/stable/index.html)\n",
    "- currently having issues with setting the variable (barcode) names to the AnnData object created from ScanPy\n",
    "- also attempting an alternative method to import data using SciPy and Pandas\n",
    "\n",
    "10-15-2020:\n",
    "- Learned how to import the data with SciPy and Pandas instead of ScanPy.\n",
    "- Continuing to create sctype class and attempting to implement SVM"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Approach #1 to importing data: ScanPy (10-13-2020)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read in the 10Xv2 data\n",
    "- The matrix.mtx, barcodes.tsv, and genes.tsv files\n",
    "- Set the names of observations and variables in the AnnData object to genes and barcodes\n",
    "\n",
    "Note: for some reason, the variable names (barcodes) isn't being labeled properly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = sc.read_mtx(filename = path + 'filtered_matrices_mex/hg19/matrix.mtx')\n",
    "var = pd.read_csv(path + 'filtered_matrices_mex/hg19/barcodes.tsv', sep = '\\t', header = None)\n",
    "obs = pd.read_csv(path + 'filtered_matrices_mex/hg19/genes.tsv', sep = '\\t', header = None)\n",
    "data.vars_names = var.iloc[:,0]\n",
    "data.obs_names = obs.iloc[:,1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Transpose the AnnData object and convert to a pd.DataFrame\n",
    "Even though expression matrices usually have genes as rows and barcodes as columns, it seems that I need to have features as the columns and each single cell set to the rows for \n",
    "\n",
    "Note: Getting a warning \"names are not unique\" for the variables (supposed to be barcodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MIR1302-10</th>\n",
       "      <th>FAM138A</th>\n",
       "      <th>OR4F5</th>\n",
       "      <th>RP11-34P13.7</th>\n",
       "      <th>RP11-34P13.8</th>\n",
       "      <th>AL627309.1</th>\n",
       "      <th>RP11-34P13.14</th>\n",
       "      <th>RP11-34P13.9</th>\n",
       "      <th>AP006222.2</th>\n",
       "      <th>RP4-669L17.10</th>\n",
       "      <th>...</th>\n",
       "      <th>KIR3DL2</th>\n",
       "      <th>AL590523.1</th>\n",
       "      <th>CT476828.1</th>\n",
       "      <th>PNRC2</th>\n",
       "      <th>SRSF10</th>\n",
       "      <th>AC145205.1</th>\n",
       "      <th>BAGE5</th>\n",
       "      <th>CU459201.1</th>\n",
       "      <th>AC002321.2</th>\n",
       "      <th>AC002321.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68574</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68575</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68576</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68577</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68578</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>68579 rows × 32738 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       MIR1302-10  FAM138A  OR4F5  RP11-34P13.7  RP11-34P13.8  AL627309.1  \\\n",
       "0             0.0      0.0    0.0           0.0           0.0         0.0   \n",
       "1             0.0      0.0    0.0           0.0           0.0         0.0   \n",
       "2             0.0      0.0    0.0           0.0           0.0         0.0   \n",
       "3             0.0      0.0    0.0           0.0           0.0         0.0   \n",
       "4             0.0      0.0    0.0           0.0           0.0         0.0   \n",
       "...           ...      ...    ...           ...           ...         ...   \n",
       "68574         0.0      0.0    0.0           0.0           0.0         0.0   \n",
       "68575         0.0      0.0    0.0           0.0           0.0         0.0   \n",
       "68576         0.0      0.0    0.0           0.0           0.0         0.0   \n",
       "68577         0.0      0.0    0.0           0.0           0.0         0.0   \n",
       "68578         0.0      0.0    0.0           0.0           0.0         0.0   \n",
       "\n",
       "       RP11-34P13.14  RP11-34P13.9  AP006222.2  RP4-669L17.10  ...  KIR3DL2  \\\n",
       "0                0.0           0.0         0.0            0.0  ...      0.0   \n",
       "1                0.0           0.0         0.0            0.0  ...      0.0   \n",
       "2                0.0           0.0         0.0            0.0  ...      0.0   \n",
       "3                0.0           0.0         0.0            0.0  ...      0.0   \n",
       "4                0.0           0.0         0.0            0.0  ...      0.0   \n",
       "...              ...           ...         ...            ...  ...      ...   \n",
       "68574            0.0           0.0         0.0            0.0  ...      0.0   \n",
       "68575            0.0           0.0         0.0            0.0  ...      0.0   \n",
       "68576            0.0           0.0         0.0            0.0  ...      0.0   \n",
       "68577            0.0           0.0         0.0            0.0  ...      0.0   \n",
       "68578            0.0           0.0         0.0            0.0  ...      0.0   \n",
       "\n",
       "       AL590523.1  CT476828.1  PNRC2  SRSF10  AC145205.1  BAGE5  CU459201.1  \\\n",
       "0             0.0         0.0    1.0     0.0         0.0    0.0         0.0   \n",
       "1             0.0         0.0    0.0     0.0         0.0    0.0         0.0   \n",
       "2             0.0         0.0    0.0     0.0         0.0    0.0         0.0   \n",
       "3             0.0         0.0    0.0     0.0         0.0    0.0         0.0   \n",
       "4             0.0         0.0    0.0     0.0         0.0    0.0         0.0   \n",
       "...           ...         ...    ...     ...         ...    ...         ...   \n",
       "68574         0.0         0.0    0.0     0.0         0.0    0.0         0.0   \n",
       "68575         0.0         0.0    0.0     0.0         0.0    0.0         0.0   \n",
       "68576         0.0         0.0    0.0     0.0         0.0    0.0         0.0   \n",
       "68577         0.0         0.0    0.0     0.0         0.0    0.0         0.0   \n",
       "68578         0.0         0.0    1.0     0.0         0.0    0.0         0.0   \n",
       "\n",
       "       AC002321.2  AC002321.1  \n",
       "0             0.0         0.0  \n",
       "1             0.0         0.0  \n",
       "2             0.0         0.0  \n",
       "3             0.0         0.0  \n",
       "4             0.0         0.0  \n",
       "...           ...         ...  \n",
       "68574         0.0         0.0  \n",
       "68575         0.0         0.0  \n",
       "68576         0.0         0.0  \n",
       "68577         0.0         0.0  \n",
       "68578         0.0         0.0  \n",
       "\n",
       "[68579 rows x 32738 columns]"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_df = data.T.to_df()\n",
    "data_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Approach #2 to importing data: SciPy and Pandas (10-15-2020)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import scipy.io as io"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import 10Xv2 data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = '/Users/leealj/py_projects/biof509_final/zheng68k/filtered_matrices_mex/hg19/'\n",
    "genes = pd.read_csv(path + 'genes.tsv', sep = '\\t', header = None)\n",
    "barcodes = pd.read_csv(path + 'barcodes.tsv', sep = '\\t', header = None)\n",
    "expression = io.mmread(path + 'matrix.mtx') \n",
    "data = pd.DataFrame.sparse.from_spmatrix(data = expression, index = genes, columns = barcodes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import the 'correct' annotations made by the original authors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TSNE.1</th>\n",
       "      <th>TSNE.2</th>\n",
       "      <th>barcodes</th>\n",
       "      <th>celltype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7.565540</td>\n",
       "      <td>0.441370</td>\n",
       "      <td>AAACATACACCCAA-1</td>\n",
       "      <td>CD8+ Cytotoxic T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.552626</td>\n",
       "      <td>-25.786672</td>\n",
       "      <td>AAACATACCCCTCA-1</td>\n",
       "      <td>CD8+/CD45RA+ Naive Cytotoxic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-5.771831</td>\n",
       "      <td>11.830846</td>\n",
       "      <td>AAACATACCGGAGA-1</td>\n",
       "      <td>CD4+/CD45RO+ Memory</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.762556</td>\n",
       "      <td>25.979346</td>\n",
       "      <td>AAACATACTAACCG-1</td>\n",
       "      <td>CD19+ B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-16.793856</td>\n",
       "      <td>-16.589970</td>\n",
       "      <td>AAACATACTCTTCA-1</td>\n",
       "      <td>CD4+/CD25 T Reg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68574</th>\n",
       "      <td>1.430476</td>\n",
       "      <td>-23.815174</td>\n",
       "      <td>TTTGCATGAGCCTA-8</td>\n",
       "      <td>CD8+ Cytotoxic T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68575</th>\n",
       "      <td>3.120762</td>\n",
       "      <td>-19.108131</td>\n",
       "      <td>TTTGCATGCTAGCA-8</td>\n",
       "      <td>CD8+/CD45RA+ Naive Cytotoxic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68576</th>\n",
       "      <td>13.526124</td>\n",
       "      <td>-1.559099</td>\n",
       "      <td>TTTGCATGCTGCAA-8</td>\n",
       "      <td>CD8+ Cytotoxic T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68577</th>\n",
       "      <td>11.646083</td>\n",
       "      <td>-3.386890</td>\n",
       "      <td>TTTGCATGGCTCCT-8</td>\n",
       "      <td>CD8+ Cytotoxic T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68578</th>\n",
       "      <td>-5.678085</td>\n",
       "      <td>-1.723097</td>\n",
       "      <td>TTTGCATGTGGTAC-8</td>\n",
       "      <td>CD8+ Cytotoxic T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>68579 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          TSNE.1     TSNE.2          barcodes                      celltype\n",
       "0       7.565540   0.441370  AAACATACACCCAA-1              CD8+ Cytotoxic T\n",
       "1       2.552626 -25.786672  AAACATACCCCTCA-1  CD8+/CD45RA+ Naive Cytotoxic\n",
       "2      -5.771831  11.830846  AAACATACCGGAGA-1           CD4+/CD45RO+ Memory\n",
       "3       1.762556  25.979346  AAACATACTAACCG-1                       CD19+ B\n",
       "4     -16.793856 -16.589970  AAACATACTCTTCA-1               CD4+/CD25 T Reg\n",
       "...          ...        ...               ...                           ...\n",
       "68574   1.430476 -23.815174  TTTGCATGAGCCTA-8              CD8+ Cytotoxic T\n",
       "68575   3.120762 -19.108131  TTTGCATGCTAGCA-8  CD8+/CD45RA+ Naive Cytotoxic\n",
       "68576  13.526124  -1.559099  TTTGCATGCTGCAA-8              CD8+ Cytotoxic T\n",
       "68577  11.646083  -3.386890  TTTGCATGGCTCCT-8              CD8+ Cytotoxic T\n",
       "68578  -5.678085  -1.723097  TTTGCATGTGGTAC-8              CD8+ Cytotoxic T\n",
       "\n",
       "[68579 rows x 4 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = '/Users/leealj/py_projects/biof509_final/zheng68k/'\n",
    "anno = pd.read_csv(path + '68k_pbmc_barcodes_annotation.tsv', sep = '\\t')\n",
    "anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.nanmax(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.max(expression.data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
